`cv::Mat_<T>::copyTo` can no longer be used on empty matrices

### System Information

After updating to OpenCV 4.13.0 one of my unit tests started failing with an OpenCV exception while trying to invoke `copyTo` on empty matrices of fixed type. In general, `copyTo` on empty matrices should be a no op and always succeed.

Bisecting the issue points to 64bbaa9f41be3e96a883cdbc11635eb84f18140c (#27972) as the commit that introduced the regression.

### Detailed description

`cv::Mat::copyTo` can no longer be invoked on empty matrices of fixed type. Trying to do so fails with the following exception:
```console
OpenCV(4.13.0) Error: Assertion failed (!fixedType() || ((Mat*)obj)->type() == mtype) in create, file /home/sergiu/Projects/opencv/modules/core/src/matrix_wrap.cpp, line 1321
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.13.0) /home/sergiu/Projects/opencv/modules/core/src/matrix_wrap.cpp:1321: error: (-215:Assertion failed) !fixedType() || ((Mat*)obj)->type() == mtype in function 'create'
```

### Steps to reproduce

```cpp
#include <opencv2/core.hpp>

int main()
{
    cv::Mat_<double> a;
    cv::Mat_<double> b;
    a.copyTo(b);
}
```

### Issue submission checklist

- [x] I report the issue, it's not a question
- [ ] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [x] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/core/src/matrix.cpp
+++ b/modules/core/src/matrix.cpp
@@ -553,7 +553,7 @@ void Mat::release()
     for(int i = 0; i < dims; i++)
         size.p[i] = 0;
 #ifdef _DEBUG
-    flags = MAGIC_VAL;
+    flags = (flags & CV_MAT_TYPE_MASK) | MAGIC_VAL;
     dims = rows = cols = 0;
     if(step.p != step.buf)
     {

--- a/modules/core/test/test_mat.cpp
+++ b/modules/core/test/test_mat.cpp
@@ -1394,6 +1394,25 @@ TEST(Core_Mat, copyToConvertTo_Empty)
     ASSERT_EQ(C.type(), CV_32SC2);
 }
 
+// Regression test for https://github.com/opencv/opencv/issues/28343
+// copyTo on empty fixed-type matrices should be a no-op and succeed
+template <typename T> class Core_Mat_copyTo : public testing::Test {};
+TYPED_TEST_CASE_P(Core_Mat_copyTo);
+
+TYPED_TEST_P(Core_Mat_copyTo, EmptyFixedType)
+{
+    cv::Mat_<TypeParam> a;
+    cv::Mat_<TypeParam> b;
+    EXPECT_NO_THROW(a.copyTo(b));
+    EXPECT_TRUE(b.empty());
+    // Verify type is still consistent after copyTo
+    EXPECT_EQ(b.type(), cv::traits::Type<TypeParam>::value);
+}
+
+REGISTER_TYPED_TEST_CASE_P(Core_Mat_copyTo, EmptyFixedType);
+typedef ::testing::Types<uchar, schar, ushort, short, int, float, double> AllMatDepths;
+INSTANTIATE_TYPED_TEST_CASE_P(CopyToTest, Core_Mat_copyTo, AllMatDepths);
+
 TEST(Core_Mat, copyNx1ToVector)
 {
     cv::Mat_<uchar> src(5, 1);

